### PR TITLE
[cherry-pick] Fix and enhance beam_search_op and beam_searc_decode_op

### PR DESCRIPTION
**Pull request by guoshengCS**
[cherry-pick] Fix and enhance beam_search_op and beam_searc_decode_op. Related #11238

### CODE REVIEW

**Comment by panyx0718** [comment]
sorry, did I miss this PR?

**Comment by kuke** [comment]
@panyx0718 Yes, this PR needs to merge into v0.14.0

**Comment by kuke** [comment]
@guoshengCS

File: python/paddle/fluid/tests/unittests/test_beam_search_op.py [hunk 1]
```diff
@@ -47,19 +51,31 @@ def test_run(self):
...
+        tensor = create_tensor(self.scope, 'pre_scores', np_data)
 
     def _create_ids(self):
-        self.lod = [[1, 3], [1, 1, 1, 1]]
+        self.lod = [[0, 2, 4], [0, 1, 2, 3, 4]]
         np_data = np.array(
             [[4, 2, 5], [2, 1, 3], [3, 5, 2], [8, 2, 1]], dtype='int64')
         tensor = create_tensor(self.scope, "ids", np_data)
-        tensor.set_recursive_sequence_lengths(self.lod)
+        tensor.set_lod(self.lod)
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_beam_search_op.py:78]: why is this changed from length to offset? Is there a bug in length-based lod?
@guoshengCS
